File: local-gateway/routers/habits.py
"""
习惯养成路由
POST   /api/habits — 创建习惯
GET    /api/habits — 习惯列表
GET    /api/habits/{habit_id} — 习惯详情
POST   /api/habits/{habit_id}/checkin — 习惯打卡
GET    /api/habits/{habit_id}/stats — 习惯统计
DELETE /api/habits/{habit_id} — 删除习惯
"""
from fastapi import APIRouter

from models.schemas import BaseModel, Field
from services import task_service

router = APIRouter(prefix="/habits", tags=["habits"])


class HabitCreateRequest(BaseModel):
    name: str = Field(..., description="习惯名称")
    description: str = Field("", description="习惯描述")
    frequency: str = Field("daily", description="频率: daily/weekly/monthly")
    target_count: int = Field(1, ge=1, description="目标次数")
    reminder_time: str = Field(None, description="提醒时间 (HH:MM)")
    color: str = Field("#27ae60", description="习惯颜色")


class CheckinRequest(BaseModel):
    count: int = Field(1, ge=1, description="打卡次数")
    note: str = Field("", description="打卡备注")


@router.post("/")
async def create_habit(request: HabitCreateRequest):
    """创建习惯"""
    result = await task_service.create_habit(
        name=request.name,
        description=request.description,
        frequency=request.frequency,
        target_count=request.target_count,
        reminder_time=request.reminder_time,
        color=request.color,
    )
    return result


@router.get("/")
async def list_habits():
    """获取所有习惯"""
    habits = await task_service.get_all_habits()
    return {"status": "success", "habits": habits}


@router.get("/{habit_id}")
async def get_habit(habit_id: str):
    """获取习惯详情（含打卡记录）"""
    habit = await task_service.get_habit(habit_id)
    if not habit:
        return {"status": "error", "message": f"习惯 {habit_id} 不存在"}
    return {"status": "success", "habit": habit}


@router.post("/{habit_id}/checkin")
async def checkin_habit(habit_id: str, request: CheckinRequest):
    """习惯打卡"""
    return await task_service.checkin_habit(habit_id, request.count, request.note)


@router.get("/{habit_id}/stats")
async def get_habit_stats(habit_id: str):
    """获取习惯统计"""
    return await task_service.get_habit_stats(habit_id)


@router.delete("/{habit_id}")
async def delete_habit(habit_id: str):
    """删除习惯"""
    return await task_service.delete_habit(habit_id)

File: local-gateway/static/js/components/habits.js
/* local-gateway/static/js/components/habits.js */
import { apiGet, apiPost, toast } from '../api.js';
import { formatTime, escapeHtml } from '../utils.js';

document.addEventListener('alpine:init', () => {
  Alpine.data('habits', () => ({
    loading: true,
    error: null,
    habits: [],
    showForm: false,
    form: { name: '', description: '', frequency: 'daily', target: 1 },
    checkingInId: null,

    async init() {
      this.$watch('$store.view.current', (val) => {
        if (val === 'habits') this.load();
      });
      if (Alpine.store('view').current === 'habits') await this.load();
    },

    async load() {
      this.loading = true;
      this.error = null;
      try {
        const data = await apiGet('/api/habits');
        const habits = data.habits || [];

        // Load streak for each habit in parallel
        this.habits = await Promise.all(habits.map(async (h) => {
          try {
            const detail = await apiGet(`/api/habits/${h.habit_id}`);
            return { ...h, streak: detail.habit?.streak || 0 };
          } catch {
            return { ...h, streak: 0 };
          }
        }));
      } catch (e) {
        this.error = e.message;
        toast('加载习惯失败', 'error');
      } finally {
        this.loading = false;
      }
    },

    async checkin(id) {
      this.checkingInId = id;
      try {
        const data = await apiPost(`/api/habits/${id}/checkin`, { count: 1, note: '' });
        if (data.status !== 'success') throw new Error(data.message || '打卡失败');
        toast('打卡成功！', 'success');

        // Refresh streak for this habit
        const habit = this.habits.find(h => h.habit_id === id);
        if (habit) {
          try {
            const detail = await apiGet(`/api/habits/${id}`);
            habit.streak = detail.habit?.streak || 0;
          } catch {}
        }
      } catch (e) {
        toast('打卡失败', 'error');
      }
      setTimeout(() => { this.checkingInId = null; }, 600);
    },

    resetForm() {
      this.form = { name: '', description: '', frequency: 'daily', target: 1 };
      this.showForm = false;
    },

    async createHabit() {
      if (!this.form.name) {
        toast('请输入习惯名称', 'error');
        return;
      }
      try {
        const data = await apiPost('/api/habits/', {
          name: this.form.name,
          description: this.form.description,
          frequency: this.form.frequency,
          target_count: this.form.target,
        });
        if (data.status !== 'success') throw new Error(data.message || '创建失败');
        toast('习惯已创建', 'success');
        this.resetForm();
        await this.load();
      } catch (e) {
        toast('创建失败', 'error');
      }
    },

    async deleteHabit(id) {
      if (!confirm('确定删除此习惯？')) return;
      this.habits = this.habits.filter(h => h.habit_id !== id);

      try {
        const resp = await fetch(`/api/habits/${id}`, { method: 'DELETE' });
        const data = await resp.json();
        if (data.status !== 'success') throw new Error(data.message);
        toast('习惯已删除', 'success');
      } catch (e) {
        toast('删除失败', 'error');
        await this.load();
      }
    },

    streakLabel(streak) {
      if (streak >= 7) return `🔥 ${streak}天`;
      if (streak >= 3) return `💪 ${streak}天`;
      return `⭐ ${streak}天`;
    },

    frequencyLabel(freq) {
      const map = { daily: '每天', weekly: '每周', monthly: '每月' };
      return map[freq] || freq;
    },

    formatTime: formatTime,
    escapeHtml: escapeHtml,
  }));
});
